### PR TITLE
Trade supports multiple routes

### PR DESCRIPTION
**Pull request by willpote**
Trade updated to support multiple routes. Change is back in a backward compatible way to avoid a major version bump: you will hit an error if you try and access the 'route' property when your Trade consists of multiple routes

I previously made a change to the SwapRouter to support multiple Trade objects being provided to `swapCallParameters`. With this change that is no longer necessary, but I left it in anyway.

### CODE REVIEW

File: src/entities/trade.ts [hunk 1]
```diff
@@ -68,9 +64,9 @@ export interface BestTradeOptions {
  */
 export class Trade<TInput extends Currency, TOutput extends Currency, TTradeType extends TradeType> {
   /**
-   * The route of the trade, i.e. which pools the trade goes through.
+   * The routes of the trade, i.e. which pools the trade goes through.
    */
-  public readonly route: Route<TInput, TOutput>
+  public readonly routes: { percent: Percent; route: Route<TInput, TOutput> }[]
```
Comment (NoahZinsmeister) [inline src/entities/trade.ts:69]: hmm, a little worried about `percent`. even though in theory a `Percent` can be arbitrarily precise, it's obviously possible to pass e.g 50% with an input amount of 101, which might lead to some undesirable rounding behavior

File: src/entities/trade.ts [hunk 2]
```diff
@@ -160,45 +171,58 @@ export class Trade<TInput extends Currency, TOutput extends Currency, TTradeType
...
+      totalOutputAmount = totalOutputAmount.add(outputAmount)
     }
 
     return new Trade({
-      route,
+      routes,
       tradeType,
-      inputAmount,
-      outputAmount
+      inputAmount: totalInputAmount,
```
Comment (NoahZinsmeister) [inline src/entities/trade.ts:224]: here's where the percent rounding issue could bite, i think

File: src/entities/trade.ts [hunk 3]
```diff
@@ -232,19 +256,42 @@ export class Trade<TInput extends Currency, TOutput extends Currency, TTradeType
...
+    invariant(
+      routes.every(({ route }) => outputAmount.currency.equals(route.output)),
+      'OUTPUT_CURRENCY_MATCH'
+    )
+
+    const totalPercent = routes
+      .map(({ percent }) => percent)
+      .reduce((total, cur) => total.add(cur), new Percent(0, 100))
+
+    invariant(totalPercent.equalTo(new Percent(100, 100)), 'TOTAL_PERCENT')
```
Comment (NoahZinsmeister) [inline src/entities/trade.ts:282]: hmm, i guess we could ensure full precision here. for e.g. exact input trades, you could just ensure that for each percent in routes, there's no remainder from multiplying.

**Review by NoahZinsmeister — COMMENTED** [review]
just did a brief first pass, seems generally sound, let me take a bit more time to sit with it

File: src/entities/trade.ts (hunk 1 above)
Comment (willpote) [reply src/entities/trade.ts]: Good call out, chatted with Moody and he proposed carrying over the remainder into the next percentage. And in total it should all net out. I implemented it in the latest revision lmk what you think

File: src/entities/trade.ts (hunk 3 above)
Comment (willpote) [reply src/entities/trade.ts:385]: I added various checks of this nature to the fromRoutes method and the constructor itself

**Comment by willpote** [comment]
I've changed the approach here, decided to remove all percentages and just use the actual amounts for each split

The routing api will handle potential loss of precision from doing splits and return amounts per route s.t. they add up correctly

File: src/entities/trade.test.ts [hunk 4]
```diff
@@ -377,15 +698,21 @@ describe('Trade', () => {
...
+        expect(
+          exactOut
+            .maximumAmountIn(new Percent(JSBI.BigInt(5), JSBI.BigInt(100)))
+            .equalTo(CurrencyAmount.fromRawAmount(token0, 16262))
+        ).toBeTruthy()
+        expect(
+          exactOut
+            .maximumAmountIn(new Percent(JSBI.BigInt(200), JSBI.BigInt(100)))
+            .equalTo(CurrencyAmount.fromRawAmount(token0, 46464))
+        ).toBeTruthy()
```
Comment (moodysalem) [inline src/entities/trade.test.ts:715]: could maybe also just define a matcher, not that it matters

File: src/entities/trade.ts [hunk 5]
```diff
@@ -33,7 +23,9 @@ export function tradeComparator<TInput extends Currency, TOutput extends Currenc
   if (a.outputAmount.equalTo(b.outputAmount)) {
     if (a.inputAmount.equalTo(b.inputAmount)) {
       // consider the number of hops since each hop costs gas
-      return a.route.tokenPath.length - b.route.tokenPath.length
+      const aHops = a.routes.map(({ route }) => route.tokenPath.length).reduce((total, cur) => total + cur, 0)
```
Comment (moodysalem) [inline src/entities/trade.ts:26]: map and then reduce can just be reduce

File: src/entities/trade.ts [hunk 6]
```diff
@@ -68,27 +64,83 @@ export interface BestTradeOptions {
...
-  public readonly route: Route<TInput, TOutput>
+  public get route(): Route<TInput, TOutput> {
+    invariant(this.routes.length == 1, 'MULTIPLE_ROUTES')
+    return this.routes[0].route
+  }
+
+  /**
+   * The routes of the trade, i.e. which routes the trade consists of.
+   */
+  public readonly routes: {
```
Comment (moodysalem) [inline src/entities/trade.ts:81]: this feels like we're expanding the word route to mean what used to be trade, can we name it something different

File: src/entities/trade.ts [hunk 7]
```diff
@@ -114,10 +166,20 @@ export class Trade<TInput extends Currency, TOutput extends Currency, TTradeType
...
-      (this._priceImpact = computePriceImpact(this.route.midPrice, this.inputAmount, this.outputAmount))
-    )
+    if (this._priceImpact) {
+      return this._priceImpact
+    }
+
+    let spotOutputAmount = CurrencyAmount.fromRawAmount(this.outputAmount.currency, 0)
+    for (const { route, inputAmount } of this.routes) {
+      const midPrice = route.midPrice
+      spotOutputAmount = spotOutputAmount.add(midPrice.quote(inputAmount))
```
Comment (moodysalem) [inline src/entities/trade.ts:176]: mmmm is this correct... i don't really have an intuition about the meaning of price impact across routes

File: src/entities/trade.ts (hunk 6 above)
Comment (willpote) [reply src/entities/trade.ts:81]: routeAmounts?

File: src/entities/trade.ts (hunk 7 above)
Comment (willpote) [reply src/entities/trade.ts:176]: I think so, and I think @NoahZinsmeister agrees?

Comment (NoahZinsmeister) [reply src/entities/trade.ts:177]: this is what i suggested, it seems consistent with how we were calculating it in the single-route case, and works out to a weighted average price impact across routes, where the weights are the quote amounts and the values are the individual price impact

**Review by NoahZinsmeister — APPROVED** [review]
LGTM

**Pull request merged**